src/comm.ts

```diff
@@ -0,0 +1,379 @@
+    return promise;
+  }
+
+  /**
+   * Register the handler for a "comm_open" message on a given targetName.
+   */
+  registerTarget(targetName: string, cb: (comm: IComm, data: any) => void): void {
+    this._targets.set(targetName, cb);
+  }
+
+  /**
+   * Send a 'comm_info_request', and return the contents of the
+   * 'comm_info_reply'.
+   */
+  commInfo(targetName?: string): Promise<ICommInfo> {
+    var contents = {};
+    if (targetName !== void 0) {
+      contents = { target_name: targetName };
+    }
+    var future = sendCommMessage(this._kernel, 'comm_info_request', contents);
+    return new Promise((resolve, reject) => {
+      future.onReply = (msg) => {
+        resolve(msg.content);
+      }
+    });
+  }
+
+  private _handleKernelMsg(kernel: IKernel, msg: IKernelMessage): void {
+    switch(msg.header.msg_type) {
+       case 'comm_open':
+         this._handleOpen(msg);
+         break;
+       case 'comm_msg':
+         this._handleMsg(msg);
+         break;
+       case 'comm_close':
+         this._handleClose(msg);
+         break;
+    }
+  }
+
+  /**
+   * Handle 'comm_open' kernel message.
+   */  
+  private _handleOpen(msg: IKernelMessage): void {
+    var content = msg.content;
+    var promise = loadTarget(content.target_name, content.target_module, 
+        this._targets).then((target: (comm: IComm, data: any) => any) => {
+      var comm = new Comm(content.target_name, content.comm_id, this._kernel,
+        () => { this._unregisterComm(content.comm_id); });
+      try {
+        var response = target(comm, content.data);
+      } catch (e) {
+        comm.close();
+        this._unregisterComm(comm.commId);
+        console.error("Exception opening new comm");
+        return Promise.reject(e);
+      }
+      // Regardless of the target return value, we need to
+      // then return the comm
+      return Promise.resolve(response).then(() => { return comm; });
+    }, () => { throw Error('Could not open comm')} );
+    this._comms.set(content.comm_id, promise);
+  }
+    
+  /**
+   * Handle 'comm_close' kernel message.
+   */  
+  private _handleClose(msg: IKernelMessage): void {
+    var content = msg.content;
+    var promise = this._comms.get(content.comm_id);
+    if (!promise) {
+        console.error('Comm promise not found for comm id ' + content.comm_id);
+        return;
+    }
+    promise.then((comm) => {
+      this._unregisterComm(comm.commId);
+      try {
+        comm.close(msg.content.data);
+      } catch (e) {
+        console.log("Exception closing comm: ", e, e.stack, msg);
+      }
+    });
+  }
+
+  /**
+   * Handle 'comm_msg' kernel message.
+   */  
+  private _handleMsg(msg: IKernelMessage): void {
+    var content = msg.content;
+    var promise = this._comms.get(content.comm_id);
+    if (!promise) {
+      console.error('Comm promise not found for comm id ' + content.comm_id);
+      return;
+    }
+    var newPromise = promise.then((comm) => {
```

Comment:
We're going to need another approach to storing the mapping of comms. Running through a promise to dispatch each message will be way too expensive, but in CPU cycles and GC churn.